src/qibolab/backends.py

```diff
@@ -125,6 +125,17 @@ def execute_circuit(self, circuit, initial_state=None, nshots=1000):
                 "Hardware backend only supports circuits as initial states.",
             )
 
+        if not all(q in circuit.wire_names for q in self.platform.qubits):
+            # This should be done in qibo side
+            # raise_error(
+            #     ValueError,
+            #     "Circuit qubits do not match the platform qubits.",
+            # )
+
+            # Temporary fix: overwrite the wire names
+            circuit._wire_names = self.qubits
+
+        self.platform.wire_names = circuit.wire_names
```

Comment:
Instead of overwriting the `platform` (which is not a great idea, since it should not be mutated by `Circuit`s execution), you could just copy the gate as soon as you access them:
https://github.com/qiboteam/qibolab/blob/4e3f0c71f9533aea8b0a6291aa7d70bcac333fed/src/qibolab/compilers/compiler.py#L151
and translate its `.qubits` attribute with the `Circuit.wire_names`

I.e.:
```py
for gate in set(filter(lambda x: x is not None, moment)):
    gate = deepcopy(gate)
    gate.qubits = tuple(circuit.wire_names[q] for q in gate.qubits)
```
(not that simple, since you need to separately replace `target_qubits` and `control_qubits` - but that's the idea)

Reply:
That would work. 

Another option which does not require mutating qibo gates (even copies) in qibolab, is to refactor the compiler rules to accept `qubits` instead of whole `gate` objects. For parametrized gates we also need `parameters`, but that could be an optional argument with `None` default. Then we would just do the mapping from qibo integer qubits to platform qubits using the wire names in a signle place in the compiler body, probably around here:
https://github.com/qiboteam/qibolab/blob/4e3f0c71f9533aea8b0a6291aa7d70bcac333fed/src/qibolab/compilers/compiler.py#L107

right before the rules are called.

Reply:
Yes, that's a possibly even better option, since it would make rules depending on more minimal objects, and you depend less on manipulations of more complex ones (which may be harder to maintain, since you'd be sensitive to upstream changes). And it'd be more uniform to the 0.2 handling.